core/src/apps/cardano/sign_tx.py

```diff
@@ -171,68 +147,39 @@ def _process_outputs(self):
     def _build_witnesses(self, tx_aux_hash: str):
         witnesses = []
         for input in self.inputs:
-            _, node = derive_address_and_node(self.keychain, input.address_n)
-            message = (
-                b"\x01" + cbor.encode(self.protocol_magic) + b"\x58\x20" + tx_aux_hash
-            )
-            signature = ed25519.sign_ext(
-                node.private_key(), node.private_key_ext(), message
-            )
-            extended_public_key = (
-                remove_ed25519_prefix(node.public_key()) + node.chain_code()
-            )
-            witnesses.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([extended_public_key, signature])),
-                ]
-            )
-
-        return witnesses
-
-    @staticmethod
-    def compute_fee(input_coins_sum: int, outgoing_coins: list, change_coins: list):
-        outgoing_coins_sum = sum(outgoing_coins)
-        change_coins_sum = sum(change_coins)
-
-        return input_coins_sum - outgoing_coins_sum - change_coins_sum
+            node = self.keychain.derive(input.address_n)
+            message = b"\x58\x20" + tx_aux_hash
+
+            signature = ed25519.sign_ext(node.private_key(), node.private_key_ext(), message)
+
+            public_key = remove_ed25519_prefix(node.public_key())
+
+            witnesses.append([public_key, signature])
+
+        return {0: witnesses}
 
     def serialise_tx(self):
 
         self._process_outputs()
 
         inputs_cbor = []
         for input in self.inputs:
-            inputs_cbor.append(
-                [
-                    (input.type or 0),
-                    cbor.Tagged(24, cbor.encode([input.prev_hash, input.prev_index])),
-                ]
-            )
-
-        inputs_cbor = cbor.IndefiniteLengthArray(inputs_cbor)
+            inputs_cbor.append([input.prev_hash, input.prev_index])
 
         outputs_cbor = []
         for index, address in enumerate(self.output_addresses):
-            outputs_cbor.append(
-                [cbor.Raw(base58.decode(address)), self.outgoing_coins[index]]
-            )
+            outputs_cbor.append([unhexlify(address), self.outgoing_coins[index]])
```

Comment:
I don't see any reason to pass around any address as a hex string when we can pass bytes. Actually, how it got itself hex-encoded in the first place?...

Reply:
Yes, this is a good question. In Byron they were sent as Base58. So I suppose they should be sent as Bech32 now? Transaction test vectors use hex strings, so that confused me again. I would however prefer not having to decode Bech32.

Reply:
I am strongly in favor of receiving addresses as bytes (no decoding of base58 or bech32). Then the interface is unified and the first four bits of the address tell us how to display it to the user (if we want to display it at all).

I have noticed another thing here: it seems that trezor reorders outputs. Ledger does no such thing and gives the control over the order of outputs in tx to the wallet (not that it has a choice). If I am right, we should confirm with @ppershing that this is the desired behavior (perhaps post a question to ledger-dev).